Verification Compound Tasks

hudson-newey opened the pull request:
# Verification Compound Tasks

## Changes

- Adds `?task=verify-and-correct` url parameter that adds a `tag-prompt` component to the verification grid that is triggered when a `false` decision is made
- Fixes typing for `tagging` endpoint to take an `AudioRecording/AudioRecordingId` instead of `AnalysisJob/AnalysisJobId`
	- API controller: https://github.com/QutEcoacoustics/baw-server/blob/master/app/controllers/taggings_controller.rb#L37
	- `routes.rb`: https://github.com/QutEcoacoustics/baw-server/blob/master/config/routes.rb#L741

## Issues

Fixes: #2383

## Visual Changes

<img width="1372" height="796" alt="image" src="https://github.com/user-attachments/assets/5cd7d8a6-7568-4628-8661-4976e2f9338d" />

_New "verify + correct tag" task behavior in annotation search form_

---

<img width="1891" height="958" alt="image" src="https://github.com/user-attachments/assets/51a3332b-033c-43f5-abfe-193a897e8e99" />

_Verification grid with new "Correct Tag" button_

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Basic tag correction implementation

hudson-newey committed:
Complete compound tasks + new decision-made events

hudson-newey committed:
Working test suite

hudson-newey committed:
Improve copy for "verify and correct tag"

[skip ci]

hudson-newey committed:
Self review

hudson-newey commented on src/app/components/annotations/pages/annotationSearchParameters.ts:
```diff
@@ -239,6 +244,24 @@ export class AnnotationSearchParameters
+    return this._taskBehavior;
+  }
+
+  public set taskBehavior(value: string) {
```
Why? `TaskBehavior` is a subset of `string`, so this doesn't narrow the type any further.

And the setter is user facing where there is no type checking, so we can't guarantee type safety by narrowing to `TaskBehavior`

hudson-newey replied:
(I wish copilot could respond to my query here)

atruskie commented on src/app/components/annotations/pages/verification/verification.component.html:
```diff
@@ -68,6 +68,23 @@ <h2>
+      (focus)="verificationGridFocused = true"
+      (blur)="verificationGridFocused = false"
+    ></oe-tag-prompt>
+  }
```
ok discussed offline. there are some visual bugs, but:

skip should come last and should not be grouped with anything.  It also probably shouldn't be conditionally defined.

hudson-newey replied:
Skip button should be fixed by upstream

https://github.com/ecoacoustics/web-components/issues/446

Squared edges are being tracked upstream in

https://github.com/ecoacoustics/web-components/issues/447

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -146,46 +172,8 @@ class VerificationComponent
-      firstValueFrom(apiRequest);
-    }
+  protected tagTextFormatter(tag: Tag): string {
+    return tag.text;
```
Can tag ever be undefined? Do we need a safe access operator?

hudson-newey replied:
Technically no. And I think we should hard-fail here (although this is up for debate and it's not a hill I'm going to die on).

The `tagTextFormatter` is used to extract text from a tag model in the typeahead input and then cast it to a compatible web component `Tag` model which has a `text` property.

Once selected in the typeahead, the web components internal `Tag` model takes over.

---

I think we should hard fail instead of showing "undefined" or empty strings as a defensive programming practice so that if a user does somehow break the text transformer, the typeahead will throw an error and not show any results which caused an error. I think not showing a result that errored is better than trying to recover from an unexpected error and potentially silently producing bad/incorrect results.

atruskie replied:
I think hard fail is the wrong choice because this will be really frustrating. 

But if we're confident it will never happen, that's fine with me

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+    const apiRequest = this.verificationApi.createOrUpdate(
+      verification,
+      subject as AudioEvent,
+      this.session.currentUser,
```
smell: shouldn't the service know what the current user is?

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+  }
+
+  private handleVerificationDecision(subjectWrapper: SubjectWrapper): void {
+    const subject = subjectWrapper.subject as Readonly<AudioEvent>;
```
Aren't our subject's `Annotation`s?

hudson-newey replied:
Probably need to change something at the `Annotation` level model here.

`Annotation` implements the `IAudioEvent` interface, but does not extend the `AudioEvent` class.
So it has some poor type checking, and is technically incorrect.

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+      return;
+    }
+
+    this.hasUnsavedChanges = true;
```
i don't know where this is unset, but the service requests are fire and forget... so?

hudson-newey replied:
Don't fully understand what you're meaning.

`hasUnsavedChanges` is a property used by our route guard that shows the "are you sure you want to navigate away, you may lose progress" alert.

Therefore, this should be set regardless of if the service request completes or not.

---

I think there's an argument for unsetting this once the request completes/fails (Is this what you're asking for)?

atruskie replied:
Yeah, isn't the result here that we'll always have unsaved changed and the page will always alert?

I think that's another bug we're tracking though...

hudson-newey replied:
The page doesn't currently alert if no decisions have been made. If any decisions have been made, the page will always alert.

This is a bit of a remanence from before users could resume verifications, so I don't think it's as needed these days.

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+        this.handleTagCorrectionDecision(subject);
+      }
+    }
+  }
```
TODO: we should be updating our annotation model after these updates.

In particular it should now have a new tagging and a set of verifications. This can wait until we start showing verifications in history

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+        this.handleVerificationDecision(subject);
+      }
+
+      if (Object.prototype.hasOwnProperty.call(change, "newTag")) {
```
maybe leave a comment here along the lines of:

```
// if a new tag is selected then a positive verification is also emitted and saved
// if a new tag is removed, we should also see the old verification get deleted

atruskie replied:
this is comment is wrong - but we should leave a comment explaining the behaviour

hudson-newey replied:
<img width="614" height="228" alt="image" src="https://github.com/user-attachments/assets/2363e97f-df63-46f1-a9ed-7d25e56ce5b2" />

atruskie replied:
add: the verification grid does not emit a separate verification. The new tag model is the decision itself: we break it up into a tagging and a verification

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+    }
+  }
+
+  private handleVerificationDecision(subjectWrapper: SubjectWrapper): void {
```
question, how are verifications removed?

hudson-newey replied:
TODO

This should have been done a while back...

hudson-newey replied:
<img width="617" height="294" alt="image" src="https://github.com/user-attachments/assets/65d0cfbb-ef12-46cf-995f-00c53f99505c" />

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+  }
+
+  private handleTagCorrectionDecision(subjectWrapper: SubjectWrapper): void {
+    const audioEvent = subjectWrapper.subject as any;
```
this could be narrowed to `Annotation`

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+    const audioEvent = subjectWrapper.subject as any;
+    const newTagId = subjectWrapper.newTag as any;
+
+    const apiRequest = this.correctTag(audioEvent, newTagId.tag.id);
```
how do we handle the removal of `newTag`

hudson-newey replied:
TODO

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+    firstValueFrom(apiRequest);
+  }
+
+  // TODO: This logic should probably be moved to a service
```
yes it should

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -246,18 +234,107 @@ class VerificationComponent
+            correctVerification,
+            audioEvent,
+            this.session.currentUser,
+          );
```
this should not be needed: the verification grid should have emitted the verification already and it should be handles by `handleVerificationDecision`

Is this just double work? or is there a bug somewhere?

atruskie replied:
this comment is wrong

hudson-newey committed:
Review feedback

hudson-newey committed:
Bump web components + handle unset states

hudson-newey committed:
Passing tests

hudson-newey committed:
Handle a bunch more cases

I can still think of an edge case where the tag corrections would break
if two people were sharing an account and performing verifications at
the same time, but I deemed that in the interest of time, I won't deal
with this edge case right now.

When I go and clean everything up, I'll handle this edge case.

hudson-newey committed:
Bump web components version to 5.1.0

hudson-newey committed:
Change order of decision buttons for hacky fix

Two skip buttons were being created.

I just remove the second skip button and re-ordered the buttons. I don't
like the semantics of it because the buttons now read right to left, but
it's better than a broken UI or not shipping the feature in time.

I suspect this is an upstream bug to do with us only checking if a skip
button exists or not in the firstUpdated hook instead of checking every
update, meaning that SPA's can init without a skip button, causing a
skip button to be added, and then an explicit skip button is added
and we don't re-check.
This problem didn't arise during SSR rendering/fresh reloads which is
why I didn't see it during dev.
